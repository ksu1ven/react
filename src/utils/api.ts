import { apiResponse } from './types';

export async function searchCards(
  searchValue: string,
  pageNumber: number,
  pageSize: number
) {
  const url = `https://stapi.co/api/v1/rest/animal/search?pageNumber=${pageNumber}&pageSize=${pageSize}`;
  try {
    const response = searchValue
      ? await fetch(url, {
          method: 'POST',
          headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
          body: `${encodeURIComponent('name')}=${encodeURIComponent(
            searchValue
          )}`,
        })
      : await fetch(url, {
          method: 'GET',
        });

    const json: apiResponse = await response.json();
    return json;
  } catch {
    throw new Error('Error with server');
  }
}
