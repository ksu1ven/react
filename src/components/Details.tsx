import React from 'react';
import { useRouter } from 'next/router';
import { updateQueryParams } from '../utils/helpFunctions';
import { Animal } from '../utils/types';

function Details(props: Record<'detailsData', Animal>) {
  const { detailsData } = props;
  const router = useRouter();
  return (
    <aside
      className="relative w-1/3 bg-lime-700 text-white flex flex-col pt-40 items-center gap-6"
      data-testid="details"
    >
      <button
        className="absolute top-14 right-14 flex items-center justify-center w-16 h-16 bg-lime-300 text-4xl p-3 rounded-full"
        data-testid="cross"
        onClick={() => {
          const newParams = updateQueryParams(router.query, 'details', '');
          router.push(newParams.toString() ? '?' + newParams : '');
        }}
      >
        X
      </button>

      <div>
        <h1
          className="text-5xl font-extrabold text-lime-300 mb-10"
          data-testid="details-h1"
        >
          {detailsData.name}
        </h1>
        <ul className="text-3xl font-extrabold">
          <li>Avian: {detailsData.avian ? 'Yes' : 'No'}</li>
          <li>Canine: {detailsData.canine ? 'Yes' : 'No'}</li>
          <li>EarthAnimal: {detailsData.earthAnimal ? 'Yes' : 'No'}</li>
          <li>EarthInsect: {detailsData.earthInsect ? 'Yes' : 'No'}</li>
          <li>Feline: {detailsData.feline ? 'Yes' : 'No'}</li>
        </ul>
      </div>
    </aside>
  );
}

export default Details;
