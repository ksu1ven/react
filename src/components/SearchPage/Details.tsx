import { useSearchParams, useLoaderData } from 'react-router-dom';
import Loader from '../Loader';
import { Animal, apiResponse } from '../../utils/types';
import { useState } from 'react';
import { updateQueryParams } from '../../utils/helpFunctions';

function Details() {
  const [params, setParams] = useSearchParams();
  const [isLoading] = useState(false);
  const animal = useLoaderData() as Animal | null;
  console.log(animal);

  return params.has('details') ? (
    <aside className="relative w-1/3 bg-lime-700 text-white flex flex-col pt-40 items-center gap-6">
      {!isLoading ? (
        <>
          <button
            className="absolute top-14 right-14 flex items-center justify-center w-16 h-16 bg-lime-300 text-4xl p-3 rounded-full"
            onClick={() => {
              setParams(updateQueryParams(params, 'details', ''));
            }}
          >
            X
          </button>
          <h1 className="text-5xl font-extrabold text-lime-300">
            {animal?.name}
          </h1>
          <ul className="text-3xl font-extrabold">
            <li>Avian: {animal?.avian ? 'Yes' : 'No'}</li>
            <li>Canine: {animal?.canine ? 'Yes' : 'No'}</li>
            <li>EarthAnimal: {animal?.earthAnimal ? 'Yes' : 'No'}</li>
            <li>EarthInsect: {animal?.earthInsect ? 'Yes' : 'No'}</li>
            <li>Feline: {animal?.feline ? 'Yes' : 'No'}</li>
          </ul>
        </>
      ) : (
        <Loader />
      )}
    </aside>
  ) : (
    ''
  );
}

export async function loader(): Promise<Animal | null> {
  try {
    const animalName = new URL(location.href).searchParams.get('details');
    if (animalName) {
      const response = await fetch(
        'https://stapi.co/api/v1/rest/animal/search?pageNumber=0&pageSize=100',
        {
          method: 'POST',
          headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
          body: `${encodeURIComponent('name')}=${encodeURIComponent(
            animalName
          )}`,
        }
      );
      const json: apiResponse = await response.json();
      return json.animals[0];
    }
    return null;
  } catch {
    throw new Error('No such uid');
  }
}

export default Details;
