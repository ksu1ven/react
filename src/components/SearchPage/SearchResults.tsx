import { Animal } from '../../utils/types';
import pawImg from '../../assets/paw.png';
import { useContext, useEffect } from 'react';
import { SetURLSearchParams, useNavigate } from 'react-router-dom';
import { updateQueryParams } from '../../utils/helpFunctions';
import { SearchResultsContext } from './Contexts';

interface Props {
  params: URLSearchParams;
  setParams: SetURLSearchParams;
}

function SearchResults(props: Props) {
  const { params, setParams } = props;
  const searchResultsArray = useContext(SearchResultsContext);
  const navigate = useNavigate();
  function checkDescription(animal: Animal) {
    const descriptionArr: string[] = [];
    Object.entries(animal).forEach((el) => {
      if (el[1] === true) {
        descriptionArr.push(el[0]);
      }
    });
    return descriptionArr.length
      ? descriptionArr.join(', ')
      : "API don't give us description :(";
  }

  useEffect(() => {
    navigate('?' + props.params, { replace: true });
  }, [navigate, props.params]);

  return (
    <div
      className="w-1/3 m-auto mb-10 flex flex-col gap-5 "
      data-testid="search-results"
    >
      {searchResultsArray.length ? (
        searchResultsArray.map((el) => {
          const animalDescription = checkDescription(el);
          return (
            <div
              key={el.uid}
              className="flex justify-between gap-5 cursor-pointer"
              data-testid="card"
              onClick={() => {
                setParams(updateQueryParams(params, 'details', el.name));
              }}
            >
              <img src={pawImg} alt="animal picture" className="w-16 h-16" />
              <div className="grow">
                <h2 className="font-extrabold mb-1">{el.name}</h2>
                <p>Description: {animalDescription}</p>
              </div>
            </div>
          );
        })
      ) : (
        <p className="text-3xl" data-testid="no-found">
          Nothing found:(
        </p>
      )}
    </div>
  );
}

export default SearchResults;
