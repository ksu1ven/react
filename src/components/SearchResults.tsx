import React from 'react';
import { useRouter } from 'next/router';
import { updateQueryParams } from '../utils/helpFunctions';
import { Animal } from '../utils/types';
import Image from 'next/image';
import pawImg from '../assets/paw.png';

function SearchResults(props: Record<'searchResults', readonly Animal[]>) {
  const { searchResults } = props;
  const router = useRouter();

  function checkDescription(animal: Animal) {
    const descriptionArr: string[] = [];
    Object.entries(animal).forEach((el) => {
      if (el[1] === true) {
        descriptionArr.push(el[0]);
      }
    });
    return descriptionArr.length
      ? descriptionArr.join(', ')
      : "API don't give us description :(";
  }

  return (
    <div
      className="w-1/3 m-auto mb-10 flex flex-col gap-5 "
      data-testid="search-results"
    >
      {searchResults.length ? (
        searchResults.map((el) => {
          const animalDescription = checkDescription(el);
          return (
            <div
              key={el.uid}
              className="flex justify-between gap-5 cursor-pointer"
              data-testid="card"
              onClick={() => {
                router.push(
                  '?' + updateQueryParams(router.query, 'details', el.name)
                );
              }}
            >
              <Image
                src={pawImg}
                alt="animal picture"
                className="w-16 h-16"
                width={100}
                height={100}
              />
              <div className="grow">
                <h2 className="font-extrabold mb-1">{el.name}</h2>
                <p>Description: {animalDescription}</p>
              </div>
            </div>
          );
        })
      ) : (
        <p className="text-3xl" data-testid="no-found">
          Nothing found:(
        </p>
      )}
    </div>
  );
}

export default SearchResults;
