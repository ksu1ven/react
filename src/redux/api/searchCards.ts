import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/query/react';
import { apiResponse } from '../../utils/types';

type Params = {
  pageNumber: number;
  pageSize: number;
  searchValue?: string;
};

export const cardsApi = createApi({
  reducerPath: 'cardsApi',
  baseQuery: fetchBaseQuery({
    baseUrl: 'https://stapi.co/api/v1/rest/animal/search',
  }),
  endpoints: (build) => ({
    searchByValue: build.mutation<apiResponse, Params>({
      query: (params) => ({
        url: `?pageNumber=${params.pageNumber}&pageSize=${params.pageSize}`,
        method: 'POST',
        headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
        body: `${encodeURIComponent('name')}=${encodeURIComponent(
          params.searchValue || ''
        )}`,
      }),
    }),
  }),
});

export const { useSearchByValueMutation } = cardsApi;
