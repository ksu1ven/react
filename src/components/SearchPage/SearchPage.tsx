import { useState } from 'react';
import { useSearchParams, Outlet } from 'react-router-dom';
import { useSelector } from 'react-redux';
import type { RootState } from '../../redux/store/store';
import SearchForm from './SearchForm';
import SearchResults from './SearchResults';
import Pagination from './Pagination';
import Loader from '../Loader';
import ErrorBoundary from '../ErrorBoundary';
import { Animal } from '../../utils/types';
import { updateQueryParams } from '../../utils/helpFunctions';
import SelectLimit from './Select';
import { SearchResultsContext } from './Contexts';

function SearchPage() {
  const [loading] = useState(false);
  const { totalPages } = useSelector((state: RootState) => state.pagination);

  const [params, setParams] = useSearchParams();

  const [searchResultsArray] = useState<Readonly<Animal[]>>([]);

  const [errorOccured] = useState(false);

  // useEffect(() => {
  //   setLoading(true);
  //   searchCards(searchValue, pageNumber, pageSize)
  //     .then((json: apiResponse) => {
  //       setTotalPages(json.page.totalPages);
  //       setLoading(false);
  //       setSearchResultsArray(json.animals);
  //     })
  //     .catch(() => setErrorOccured(true));
  // }, [searchValue, pageNumber, pageSize]);

  if (errorOccured) {
    throw new Error("Hello, I'm Error with server!");
  }
  return (
    <ErrorBoundary>
      <main className="relative min-h-screen flex flex-col grow">
        <section className="bg-lime-200 py-10">
          <SearchForm params={params} setParams={setParams} />
        </section>
        <section className="search-results grow">
          <SelectLimit params={params} setParams={setParams} />
          {!loading && (
            <>
              <SearchResultsContext.Provider value={searchResultsArray}>
                <SearchResults params={params} setParams={setParams} />
              </SearchResultsContext.Provider>

              {totalPages && (
                <Pagination params={params} setParams={setParams} />
              )}
            </>
          )}
        </section>
        {loading && <Loader />}
      </main>
      {params.has('details') && (
        <div
          className="fixed w-screen h-full bg-slate-500/70"
          onClick={() => {
            if (params.has('details'))
              setParams(updateQueryParams(params, 'details', ''));
          }}
        ></div>
      )}
      <Outlet />
    </ErrorBoundary>
  );
}

export default SearchPage;
