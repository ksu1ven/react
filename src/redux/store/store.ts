import { configureStore } from '@reduxjs/toolkit';
import paginationSlice from '../features/paginationSlice';
import limitSlice from '../features/limitSlice';
import searchSlice from '../features/searchSlice';
import loaderSlice from '../features/loaderSlice';
import resultsSlice from '../features/resultsSlice';
import { cardsApi } from '../api/searchCards';

export const store = configureStore({
  reducer: {
    pagination: paginationSlice,
    limit: limitSlice,
    search: searchSlice,
    loader: loaderSlice,
    results: resultsSlice,
    [cardsApi.reducerPath]: cardsApi.reducer,
  },
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware().concat(cardsApi.middleware),
});

export type RootState = ReturnType<typeof store.getState>;

export type AppDispatch = typeof store.dispatch;
