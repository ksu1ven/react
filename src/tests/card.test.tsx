import { describe, expect, it, vi } from 'vitest';
import '@testing-library/jest-dom';
import { Provider } from 'react-redux';
import { store } from '../redux/store/store';
import { fireEvent, render, screen } from '@testing-library/react';
import {
  createMemoryRouter,
  RouterProvider,
  MemoryRouter,
  createRoutesFromElements,
  Route,
} from 'react-router-dom';
import Details from '../components/SearchPage/Details';
import SearchPage from '../components/SearchPage/SearchPage';
import SearchResults from '../components/SearchPage/SearchResults';
import { createCardsListResponseMock } from './mocks';
import * as mockSearchCards from '../redux/api/searchCards';

describe('Tests for the Card component', () => {
  const routes = createRoutesFromElements(
    <Route
      path="/"
      element={
        <Provider store={store}>
          <SearchPage />
        </Provider>
      }
    >
      <Route
        index
        element={
          <Provider store={store}>
            <Details />
          </Provider>
        }
      />
    </Route>
  );

  it('The card component renders the relevant card data', async () => {
    render(
      <MemoryRouter>
        <Provider store={store}>
          <SearchResults
            params={new URLSearchParams('')}
            setParams={vi.fn()}
            searchResults={createCardsListResponseMock(7, 10, false)}
          />
        </Provider>
      </MemoryRouter>
    );
    const card = (await screen.findAllByTestId('card'))[0];
    expect(card).toMatchSnapshot();
    expect(screen.getAllByAltText('animal picture')[0]).toBeTruthy();
    expect(screen.getAllByRole('heading')[3]).toHaveTextContent('testCard-3');
    expect(screen.queryByText('Some-text')).toBeFalsy();
    expect(screen.getAllByText('Description: avian')[0]).toBeTruthy();
  });

  it('Clicking on a card opens a detailed card component', async () => {
    const router = createMemoryRouter(routes);
    render(<RouterProvider router={router} />);
    const card = (await screen.findAllByTestId('card'))[0];
    expect(card).toMatchSnapshot();
    expect(screen.queryByTestId('details')).not.toBeInTheDocument();
    fireEvent.click(card);
    expect(router.state.location.search).toBe('?details=testCard-0');
    const details = await screen.findByTestId('details');
    expect(details).toMatchSnapshot();
    expect(details).toBeInTheDocument();
  });

  it('Clicking triggers an additional API call to fetch detailed information.', async () => {
    const router = createMemoryRouter(routes);
    render(<RouterProvider router={router} />);
    const card = (await screen.findAllByTestId('card'))[0];
    expect(screen.queryByTestId('details')).not.toBeInTheDocument();
    expect(card).toMatchSnapshot();

    const mockedUseSearchByValueMutation = vi.spyOn(
      mockSearchCards,
      'useSearchByValueMutation'
    );

    fireEvent.click(card);

    expect(mockedUseSearchByValueMutation).toHaveBeenCalled();
  });
});
