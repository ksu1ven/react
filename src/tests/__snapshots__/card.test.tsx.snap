// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Tests for the Card component > Clicking on a card opens a detailed card component 1`] = `
<div
  class="flex justify-between gap-5 cursor-pointer"
  data-testid="card"
>
  <img
    alt="animal picture"
    class="w-16 h-16"
    src="/src/assets/paw.png"
  />
  <div
    class="grow"
  >
    <h2
      class="font-extrabold mb-1"
    >
      testCard-0
    </h2>
    <p>
      Description: 
      avian
    </p>
  </div>
</div>
`;

exports[`Tests for the Card component > Clicking on a card opens a detailed card component 2`] = `
<aside
  class="relative w-1/3 bg-lime-700 text-white flex flex-col pt-40 items-center gap-6"
  data-testid="details"
>
  <button
    class="absolute top-14 right-14 flex items-center justify-center w-16 h-16 bg-lime-300 text-4xl p-3 rounded-full"
    data-testid="cross"
  >
    X
  </button>
  <span
    class="loader"
    data-testid="loader"
  />
</aside>
`;

exports[`Tests for the Card component > Clicking triggers an additional API call to fetch detailed information. 1`] = `
<div
  class="flex justify-between gap-5 cursor-pointer"
  data-testid="card"
>
  <img
    alt="animal picture"
    class="w-16 h-16"
    src="/src/assets/paw.png"
  />
  <div
    class="grow"
  >
    <h2
      class="font-extrabold mb-1"
    >
      testCard-0
    </h2>
    <p>
      Description: 
      avian
    </p>
  </div>
</div>
`;

exports[`Tests for the Card component > The card component renders the relevant card data 1`] = `
<div
  class="flex justify-between gap-5 cursor-pointer"
  data-testid="card"
>
  <img
    alt="animal picture"
    class="w-16 h-16"
    src="/src/assets/paw.png"
  />
  <div
    class="grow"
  >
    <h2
      class="font-extrabold mb-1"
    >
      testCard-0
    </h2>
    <p>
      Description: 
      avian
    </p>
  </div>
</div>
`;
